File: webapp/tools/index_sqlite.py
import os
from sqlalchemy import create_engine
from sqlalchemy.engine import Engine


def tenant_db_path(id: int) -> str:
    """テナントDBのパスを返す"""
    tenant_db_dir = os.getenv("ISUCON_TENANT_DB_DIR", "../tenant_db")
    return tenant_db_dir + f"/{id}.db"


def connect_to_tenant_db(id: int) -> Engine:
    """テナントDBに接続する"""
    path = tenant_db_path(id)
    print(path)
    engine = create_engine(f"sqlite:///{path}")
    return engine


for i in range(1, 101):
    tenant_db = connect_to_tenant_db(i)
    tenant_db.execute("CREATE INDEX tenant_id_idx_competition ON competition (tenant_id)")
    tenant_db.execute("CREATE INDEX tenant_id_idx_player ON player (tenant_id)")
    tenant_db.execute("CREATE INDEX score_idx ON player_score (tenant_id, competition_id, player_id)")

File: webapp/node/src/index_sqlite.ts
import path from 'path'
import sqlite3 from 'sqlite3'
import { open, Database } from 'sqlite'

// 環境変数を取得する、なければデフォルト値を返す
function getEnv(key: string, defaultValue: string): string {
    const val = process.env[key]
    if (val !== undefined) {
      return val
    }

    return defaultValue
  }


// テナントDBのパスを返す
function tenantDBPath(id: number): string {
    const tenantDBDir = getEnv('ISUCON_TENANT_DB_DIR', '../tenant_db')
    return path.join(tenantDBDir, `${id.toString()}.db`)
}

// テナントDBに接続する
async function connectToTenantDB(id: number): Promise<Database> {
    const p = tenantDBPath(id)
    let db: Database
    try {
      db = await open({
        filename: p,
        driver: sqlite3.Database,
      })
      db.configure('busyTimeout', 5000)
    } catch (error) {
      throw new Error(`failed to open tenant DB: ${error}`)
    }

    return db
}

for (let i = 1; i <= 100; i++) {
    const tenantDB = await connectToTenantDB(i)
    tenantDB.run('CREATE INDEX IF NOT EXISTS `tenant_id_idx` ON `competition` (`tenant_id`)')
    tenantDB.run('CREATE INDEX IF NOT EXISTS `tenant_id_idx` ON `player` (`tenant_id`)')
    tenantDB.run('CREATE INDEX IF NOT EXISTS `score_idx` ON `player` (`tenant_id`, `competition_id`, `player_id`)')
    tenantDB.close()
}